### ISSUE TITLE
problem connecting over Tor

### ISSUE THREAD

**Issue by islandtrail** [AsamK/signal-cli#614]
Hello everyone,

I need some help analysing a bug I've run into, and this seemed like a good place to ask - hope that's OK.

I'm trying to use signal-cli over Tor.

Unlike #261, I **was able to successfully register and verify** an account on the Signal servers over Tor (after jumping through the hoops of all those annoying reCAPTCHAs that OWS make you do if you are connecting from a Tor exit).

I was also able to issue a setPin command, and the first couple of times I issued a listIdentities command **it worked fine** (listed the fingerprint). So I know that **it is possible**.

However, this very quickly changed, and ever since then I have been **unable to run the listIdentities or receive commands**.

If I try:
`torsocks ./signal-cli --verbose listIdentities`
The error I now get is:
```
1620513792 WARNING torsocks[859]: [syscall] Unsupported syscall number 229. Denying the call (in tsocks_syscall() at syscall.c:605)
1620513795 PERROR torsocks[859]: socks5 libc connect: Invalid argument (in socks5_connect() at socks5.c:202)
Error while checking account +447533408378: java.net.ConnectException: Failed to connect to textsecure-service.whispersystems.org/54.81.92.106:443
```

The first part of this seems to be a common error that occurs when using torsocks, which can usually be safely ignored.

The error seems to be the same when testing with a different Tor circuit/exit node (I have tried about 10 different ones). All that changes is that textsecure-service.whispersystems.org resolves to a different IP (it seems to be hosted on AWS so I guess they send you to a different IP depending on which IP they see you connecting from).

However, **the actual TCP connection itself doesn't seem to be the problem**. If I test with netcat or curl, it does connect, even if I use the exact same Tor circuit/exit node that claims it 'Failed to connect' in signal-cli.

For example:
`torsocks -u foo -p bar nc -v textsecure-service.whispersystems.org 443`
...gives output:
```
Connection to textsecure-service.whispersystems.org 443 port [tcp/https] succeeded!
GET / HTTP/1.1
Host: textsecure-service.whispersystems.org

HTTP/1.1 400 Bad Request
Server: awselb/2.0
Date: Sat, 08 May 2021 22:59:26 GMT
Content-Type: text/html
Content-Length: 220
Connection: close

<html>
<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>
<body>
...
```

While:
`torsocks -u foo -p bar curl https://textsecure-service.whispersystems.org`
...gives output:
```
curl: (60) SSL certificate problem: self signed certificate in certificate chain
...
```
(because the Signal servers use a self-signed cert)

So if I try instead:
`torsocks -u foo -p bar curl -k https://textsecure-service.whispersystems.org`
...I get:
`{"code":404,"message":"HTTP 404 Not Found"}`

And yet:
`torsocks -u foo -p bar ./signal-cli --verbose listIdentities`
...still gives me:
```
1620513792 WARNING torsocks[859]: [syscall] Unsupported syscall number 229. Denying the call (in tsocks_syscall() at syscall.c:605)
1620513795 PERROR torsocks[859]: socks5 libc connect: Invalid argument (in socks5_connect() at socks5.c:202)
Error while checking account +447533408378: java.net.ConnectException: Failed to connect to textsecure-service.whispersystems.org/54.81.92.106:443
```

What can I do to get more detail about what is going on behind the scenes with signal-cli's attempt to connect? The `--verbose` flag doesn't seem to be adding any extra detail, and I really need a way to dig in deeper and discover what the problem is.

Alternatively, can I emulate the signal-cli commands myself using curl? I'm assuming it would involve all sorts of complicated crypto that would be tricky to do by hand, but maybe there is a way.

Is it possible the OWS servers are throwing more CAPTCHAs at me? Or is that something they only use during registration? They do seem to have a very Tor-unfriendly approach.

Surely I can't be the only one using signal-cli over Tor? How do I work around this problem?

Any pointers appreciated!

**Comment by exquo** [comment]
This might be `torsocks` not handling some networking call, rather than signal servers blocking connections coming from Tor. One thing to try is to take `torsocks` out of the loop: try using signal-cli in e.g. [Whonix](https://www.whonix.org/). 

A longer shot: wrap the whole interactive shell in `torsocks` and issue the `signal-cli` command from there:

	torsocks bash
	bash> signal-cli

Updating to the latest `master` version of `torsocks` might be worth a try too.

For diagnostics, a stack trace will print out lots of stuff, some of which might be helpful ([This Unix.StackExchange answer](https://unix.stackexchange.com/questions/375387/how-to-trace-networking-activity-of-a-command/375402#375402) mentions a `trace=network` argument)

> Is it possible the OWS servers are throwing more CAPTCHAs at me? Or is that something they only use during registration? 

My guess is they only do it during registration.

> They do seem to have a very Tor-unfriendly approach.

Hard to blame them, given the amount of abuse coming from the Tor network (don't get me wrong, I run an exit relay myself (and receive abuse complaints often enough)). If registration is the only hassle, they are not making things any harder for Tor users than most other high-traffic-volume internet services.

**Comment by islandtrail** [comment]
Solved it, though I now feel a bit foolish.

Short version: I had forgotten to:
`export JAVA_TOOL_OPTIONS="-Djava.net.preferIPv4Stack=true"`
before running the torsocks command.

Long version:

In theory, I already knew that in order to get torsocks to work with a Java program like signal-cli I needed to set that environment variable. In fact I had done so in the past, which must have been why it was working to begin with!

I can only assume now that at some point, I must have had several terminals open and simply switched to using a different one without realising I needed to set that variable again.

Then, when I tried again on other occasions (in case it was something being throttled or blocked by the Signal servers), I must have forgotten **again** about the necessity of setting that variable!

My saga of figuring it out:

So when I read your suggestion (to take torsocks out of the equation) I thought: "but it HAS worked with torsocks at some point in the past, so that wouldn't make sense..".

Still I tried it anyway - I set up a different way of running signal-cli through Tor and it worked!

Which left me with the mystery of why torsocks wasn't working. I learnt how to use strace on it, as you suggested, though most of the output went over my head. As I was combing through that output, I noticed a line which began:
`socket(AF_INET6 ...`
....and the penny suddenly dropped!

Some simple testing quickly showed that everything worked fine when I set that environment variable, and broke whenever I hadn't.

Thanks for your patience with this!

It's notable that I was so quick to jump to blaming Tor-unfriendly server-side policies from Signal. I guess my trust for them is pretty low because of some of their choices/priorities.

As for the CAPTCHAs in particular, it wasn't the hassle of having to do a CAPTCHA that I minded so much, but having to allow javascript in order to do so (particularly having to allow javascript from google). I would have liked to see a more thoughtful approach from an organisation that claims to be focused on privacy.

Anyway, gripes aside, respect to you for running an exit node! I imagine it's a mostly thankless task.
